qemu-tests/src/bin/x25519_on_target/main.rs

```diff
@@ -0,0 +1,77 @@
+#![no_std]
+#![no_main]
+
+extern crate panic_semihosting;
+use cortex_m_rt::entry;
+use cortex_m_semihosting::{debug, hprintln, hprint};
+
+use wycheproof_gen::generate_data;
+
+use core::convert::TryFrom;
+use salty::constants::{SECRETKEY_SEED_LENGTH, PUBLICKEY_SERIALIZED_LENGTH};
+use salty::agreement;
+
+use wycheproof::wycheproof::*;
+
+const THE_TESTS: WycheproofTest = generate_data!("../tests/x25519_test.json", "xdh_comp_schema.json");
+
+#[entry]
+fn main () -> ! {
+
+    hprint!("running tests...\n").ok();
+
+    for testgroup in THE_TESTS.test_groups {
+        if let TestGroup::XdhComp{curve, tests} = testgroup {
+            for testcase in tests.as_ref() {
+                run_x25519_comp(&curve, &testcase);
+            }
+        }
+    }
+
+    hprintln!("done.").ok();
+
+    debug::exit(debug::EXIT_SUCCESS);
+    loop { continue; }
+}
+
+fn run_x25519_comp(_curve: &str, tc: &XdhTestVector) {
+
+    hprint!("X25519 test case {:4}: ", tc.tc_id).ok();
+
+    let private = <[u8; SECRETKEY_SEED_LENGTH]>::try_from(tc.private);
+    let public  = <[u8; PUBLICKEY_SERIALIZED_LENGTH]>::try_from(tc.public);
+    let expect  = <[u8; 32]>::try_from(tc.shared);
+    let valid: bool;
+
+    if private.is_err() || public.is_err() || expect.is_err() {
+        valid = false;
+    } else {
+        let public = agreement::PublicKey::try_from(public.unwrap());
+        if public.is_err() {
+            valid = false;
+        } else {
+            let private = agreement::SecretKey::from_seed(&private.unwrap());
+            let shared  = private.agree(&public.unwrap());
+
+            valid = shared.to_bytes() == expect.unwrap();
+        }
+    }
+
+    match tc.result {
+        ExpectedResult::Valid => if !valid {
+            hprintln!("FAIL (expected VALID, but isn't)").ok();
```

Comment:
Don't we want to fail with `EXIT_FAILURE` here and elsewhere? Fine to have a "list all the errors and keep going mode", I'd just like CI to catch regressions.

Reply:
Very valid point. I added an immediate error exit for now. It is propagated through `make` to the calling shell and should work for CI tests.
In the long term, maybe run all tests but count outcome and exit with matching result at end. More like proper Rust test cases.